### PR TITLE
doc/md: add migrate down

### PR DESCRIPTION
**Pull request by a8m**
Blog post will come up in another PR.

### CODE REVIEW

File: doc/md/versioned/down.mdx [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+<details>
+<summary>Screenshot example</summary>
+
+<div style={{textAlign: 'center'}}>
+    <img src="https://atlasgo.io/uploads/versioned/protected-flow.png" alt="Migration Approval Policy" width="100%"/>
+    <p style={{fontSize: 12}}>Migration approval policy enabled for the <code>atlas migrate down</code> command</p>
+</div>
+</details>
+
+Once set, running `atlas migrate down` will create a plan that requires approvals before it is applied:
```
Comment (masseelch) [inline doc/md/versioned/down.mdx:33]:
```suggestion
Once set, running `atlas migrate down` will create a plan that requires approval before it is applied:
```

File: doc/md/versioned/down.mdx [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+</Tabs>
+
+3\. After downgrading your database to the desired version, you can safely delete the migration file `20240305171146.sql`
+from the migration directory and then run `atlas migrate hash` to update the `atlas.sum` file.
+
+4\. After the file was deleted and the database downgraded, you can generate a new migration using the `atlas migrate diff`
+    command with the optional `--edit` flag to open the generated file in your default editor.
+
+:::note Cleaning the database
+`atlas migrate down` reverts only changes applied by the migration directory, and do not touch resources such as tables that
```
Comment (masseelch) [inline doc/md/versioned/down.mdx:164]:
```suggestion
`atlas migrate down` reverts only changes applied by the migration directory, and does not touch resources that
```

File: doc/md/versioned/down.mdx [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+exist in the schema, but not managed by migration directory. If you need to clean the entire database or a schema to its initial
```
Comment (masseelch) [inline doc/md/versioned/down.mdx:165]:
```suggestion
exist in the schema, but are not managed by migration directory. If you need to clean the entire database or a schema to its initial
```

File: doc/md/versioned/down.mdx [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+#### Using the Schema Registry
+
+If the project is connected to the [Atlas Schema Registry](/cloud/features/registry), you can simply revert to a specific
+version or a tag (e.g., GitHub commit)  using the [GitHub Atlas Action](https://github.com/ariga/atlas-action?tab=readme-ov-file#arigaatlas-actionmigratedown)
+or with the following commands:
+
+```hcl title="atlas.hcl"
+env {
+  name = atlas.env # Dynamically set.
```
Comment (masseelch) [inline doc/md/versioned/down.mdx:234]:
```suggestion
  name = atlas.env # dynamically set
```

File: doc/md/versioned/down.mdx [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+  }
+}
+```
+
+<Tabs defaultValue="mysql">
+<TabItem value="mysql" label="MySQL">
+
+```shell
+atlas migrate down \
+  --to-tag "$COMMIT" \
```
Comment (masseelch) [inline doc/md/versioned/down.mdx:246]:
```suggestion
  --to-tag "$COMMIT_SHA" \
```
I think this makes it more clear.

File: doc/md/versioned/down.mdx [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+  of the execution, when Atlas generates a down migration plan, it takes the database (+ its version) and the necessary changes
+  into account. If a transactional plan, executable as a single unit, can be created, Atlas will opt for this approach.
+  If not, Atlas reverts each applied statement one at a time, ensuring the database is not left in an unknown state in case a
+  failure occurs midway, for any reason.
+  :::
+
+### Flags and Arguments
+
+The default behavior of `atlas migrate down` is that it reverts the last applied file. However, you can pass the amount
+of migrations to revert as an argument. For instance, `atlas migrate down 2` will revert up to 2 pending migrations.
```
Comment (masseelch) [inline doc/md/versioned/down.mdx:341]:
```suggestion
The default behavior of `atlas migrate down` is that it reverts the last applied file. However, you can pass the amount
of migrations to revert as an argument, or the target version or tag as a flag. For instance, `atlas migrate down 2` will revert up to 2 pending migrations while `atlas migrate down --to-version 1234` will revert all applied migrations after version "1234".
```

File: doc/md/versioned/apply.mdx [hunk 7]
```diff
@@ -30,9 +30,9 @@ The following flags are required:
 * `--dir` the URL to the migration directory. It defaults to `file://migrations`.
 
 :::info Reading remote directories
-Users who have connected or pushed their migration directories to the Atlas Schema Registry can read the migrations state directly from
-there without needing to have them locally. For example, `atlas migrate apply --dir "atlas://app"` will apply the pending
-migrations of the `app` project based on the most recent pushed state. To see it in action, run the following:
+Users who have connected or pushed their migration directory to the Atlas Schema Registry can read the migrations state
```
Comment (hilakashai) [inline doc/md/versioned/apply.mdx:33]:
```suggestion
Users who have connected or pushed their migration directory to the Atlas Schema Registry can read the migrations' state
```

File: doc/md/versioned/down.mdx [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+id: down
+slug: /versioned/down
+title: Down Migrations
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+The `atlas migrate down` command allows reverting applied migrations. Unlike the traditional approach, where down files
+are "pre-planned," Atlas computes a migration plan based on the database state to revert applied migrations and executes
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:11]:
```suggestion
are "pre-planned", Atlas computes a migration plan based on the current state of the database. Atlas reverts previously applied migrations and executes
```

File: doc/md/versioned/down.mdx [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+them until the desired version is reached, regardless of the state of the latest applied migration — whether it succeeded,
+failed, or was partially applied and left the database in an unknown version.
+
+By default, Atlas generates and executes a set of [pre-migration checks](/versioned/checks) to ensure the computed plan
+does not introduce destructive changes or delete data. Users can review the plan and run the checks before the plan is
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:16]:
```suggestion
does not introduce destructive changes or deletes data. Users can review the plan and run the checks before the plan is
```

File: doc/md/versioned/down.mdx [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+
+![Review Required](https://atlasgo.io/uploads/versioned/down-applied.png)
+
+</TabItem>
+</Tabs>
+
+
+### Dry Run
+
+The `dry-run` option allows viewing the SQL statements planned to downgrade the database to its desired version.
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:57]:
```suggestion
The `dry-run` option allows viewing the SQL statements planned to downgrade the database to its desired version, without executing them.
```

File: doc/md/versioned/down.mdx [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+If the down migration plan contains [pre-migration checks](/versioned/checks), Atlas executes them on the database and report the results.
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:58]:
```suggestion
If the down migration plan contains [pre-migration checks](/versioned/checks), Atlas execute the checks on the database and reports the results.
```

File: doc/md/versioned/down.mdx [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+    <img src="https://atlasgo.io/uploads/versioned/down-dry-run.png" alt="Migrate Down Dry Run" width="100%"/>
+    <p style={{fontSize: 12}}>Dry-run down migration in Atlas Cloud</p>
+</div>
+</details>
+
+:::
+
+### Revert local databases
+
+
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:85]:
```suggestion
```

File: doc/md/versioned/down.mdx [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+When experimenting with generating and applying migrations locally, sometimes there is a need to revert the last applied
+migration(s), typically the last one applied. The following steps demonstrate how to revert the last applied migration,
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:87]:
```suggestion
migration or migrations. The following steps demonstrate how to revert the last applied migration,
```

File: doc/md/versioned/down.mdx (hunk 2 above)
Comment (hilakashai) [inline doc/md/versioned/down.mdx:164]:
```suggestion
`atlas migrate down` only reverts changes applied by the migration directory, and does not touch resources such as tables that
```

File: doc/md/versioned/down.mdx [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+
+</TabItem>
+</Tabs>
+
+:::
+
+### Revert real environments
+
+Atlas follows a linear migration history model, in which schema changes are "roll-forward" by default rather than rolled back.
+In practice, that means reverting a schema change, such as table or column addition, require generating a new migration version
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:219]:
```suggestion
In practice, that means reverting a schema change, such as table or column addition, requires generating a new migration version
```

File: doc/md/versioned/down.mdx [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+that undo the changes, due to the fact that the "up" and "down" might not occur consecutively as the schema evolves over time.
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:220]:
```suggestion
to undo the changes. This is due to the fact that the "up" and "down" might not occur consecutively as the schema evolves over time.
```

File: doc/md/versioned/down.mdx [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+
+However, there are cases that reverting the last applied migration(s) is necessary, such when experimenting on a staging environment
+or when the entire deployment is rolled-back and the schema changes introduced in the new version are not backward compatible for old
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:223]:
```suggestion
or when the entire deployment is rolled back and the schema changes introduced in the new version are not backward compatible for old
```

File: doc/md/versioned/down.mdx [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+versions of the application and therefore need to be reverted. To undo applied migration(s) on real environments, there are two approaches:
+
+#### Using the Schema Registry
+
+If the project is connected to the [Atlas Schema Registry](/cloud/features/registry), you can simply revert to a specific
+version or a tag (e.g., GitHub commit)  using the [GitHub Atlas Action](https://github.com/ariga/atlas-action?tab=readme-ov-file#arigaatlas-actionmigratedown)
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:229]:
```suggestion
version or a tag (e.g., GitHub commit) using the [Atlas GitHub Action](https://github.com/ariga/atlas-action?tab=readme-ov-file#arigaatlas-actionmigratedown)
```

File: doc/md/versioned/down.mdx [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+atlas migrate down \
+  --to-tag "$COMMIT" \
+  --url "clickhouse://root:pass@:9000/default" \
+  --dev-url "docker://clickhouse/23.11/dev"
+```
+
+</TabItem>
+</Tabs>
+
+If this command [requires a review](#migration-approval), a new plan will be created and Atlas waits for approval before executing it.
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:309]:
```suggestion
If this command [requires a review](#migration-approval), a new plan will be created and Atlas will wait for approval before executing it.
```

File: doc/md/versioned/down.mdx [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+  In addition, since non-transactional DDLs can cause the database to stay in a stale state in case they fail in the middle
+  of the execution, when Atlas generates a down migration plan, it takes the database (+ its version) and the necessary changes
+  into account. If a transactional plan, executable as a single unit, can be created, Atlas will opt for this approach.
+  If not, Atlas reverts each applied statement one at a time, ensuring the database is not left in an unknown state in case a
+  failure occurs midway, for any reason.
+  :::
+
+### Flags and Arguments
+
+The default behavior of `atlas migrate down` is that it reverts the last applied file. However, you can pass the amount
```
Comment (hilakashai) [inline doc/md/versioned/down.mdx:340]:
```suggestion
The default behavior of `atlas migrate down` is to revert the last applied file. However, you can pass the amount
```

**Comment by a8m** [comment]
Thanks @masseelch and @hilakashai for the review. Comments addressed

**Review by masseelch — APPROVED** [review]

**Pull request merged**